gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/common/impl/InMemoryTopology.java

```diff
@@ -0,0 +1,346 @@
+  private final Map<ConfigKeyPath, List<ConfigKeyPath>> ownImportMap = new HashMap<>();
+  private final Map<ConfigKeyPath, Collection<ConfigKeyPath>> ownImportedByMap = new HashMap<>();
+  private final Map<ConfigKeyPath, List<ConfigKeyPath>> recursiveImportMap = new HashMap<>();
+  private final Map<ConfigKeyPath, Collection<ConfigKeyPath>> recursiveImportedByMap = new HashMap<>();
+
+  public InMemoryTopology(ConfigStoreTopologyInspector fallback) {
+    this.fallback = fallback;
+  }
+
+  private void loadRawTopologyFromFallBack() {
+    // breath first search the whole topology to build ownImports map and ownImportedByMap
+    // calls to retrieve cache / set cache if not present
+    Collection<ConfigKeyPath> currentLevel = this.getChildren(SingleLinkedListConfigKeyPath.ROOT);
+
+    List<ConfigKeyPath> rootImports = this.getOwnImports(SingleLinkedListConfigKeyPath.ROOT);
+    Preconditions.checkArgument(rootImports == null || rootImports.size() == 0,
+        "Root can not import other nodes, otherwise circular dependency will happen");
+
+    while (!currentLevel.isEmpty()) {
+      Collection<ConfigKeyPath> nextLevel = new ArrayList<ConfigKeyPath>();
+      for (ConfigKeyPath configKeyPath : currentLevel) {
+        // calls to retrieve cache / set cache if not present
+        List<ConfigKeyPath> ownImports = this.getOwnImports(configKeyPath);
+        for (ConfigKeyPath p : ownImports) {
+          this.addToCollectionMapForSingleValue(this.ownImportedByMap, p, configKeyPath);
+        }
+
+        // calls to retrieve cache / set cache if not present
+        Collection<ConfigKeyPath> tmp = this.getChildren(configKeyPath);
+        nextLevel.addAll(tmp);
+      }
+      currentLevel = nextLevel;
+    }
+
+    // traversal the ownImports map to get the recursive imports map and set recursive imported by map
+    for (ConfigKeyPath configKey : this.ownImportMap.keySet()) {
+      List<ConfigKeyPath> recursiveImports = null;
+      // result may in the cache already 
+      if (this.recursiveImportMap.containsKey(configKey)) {
+        recursiveImports = this.recursiveImportMap.get(configKey);
+      } else {
+        recursiveImports = this.buildImportsRecursiveFromCache(configKey);
+        this.addToListMapForListValue(recursiveImportMap, configKey, recursiveImports);
+      }
+
+      if (recursiveImports != null) {
+        for (ConfigKeyPath p : recursiveImports) {
+          this.addToCollectionMapForSingleValue(this.recursiveImportedByMap, p, configKey);
+        }
+      }
+    }
+  }
+
+  private List<ConfigKeyPath> buildImportsRecursiveFromCache(ConfigKeyPath configKey) {
+    return buildImportsRecursiveFromCacheHelper(configKey, configKey, new ArrayList<ConfigKeyPath>());
+  }
+
+  private List<ConfigKeyPath> buildImportsRecursiveFromCacheHelper(ConfigKeyPath initialConfigKey,
+      ConfigKeyPath currentConfigKey, List<ConfigKeyPath> previous) {
```

Comment:
Used LinkedHashSet because we need to maintain the order once the circular dependency happened and print out the circular chain